### PR TITLE
Vinode truncate changes

### PR DESCRIPTION
**Pull request by GBuella**
This change is [Reviewable](https://reviewable.io/reviews/marcinslusarz/pmemfile/41)

### CODE REVIEW

**Comment by marcinslusarz** [comment]
Please remember to use "posix" instead of "core" prefix in the commit message. When I'll be ready merging it I'll squash those commits into one and fix that, but please don't forget about for future PRs.

File: src/libpmemfile-core/data.c [hunk 1]
```diff
@@ -152,8 +152,10 @@ find_block(struct pmemfile_vinode *vinode, uint64_t off)
  * vinode_destroy_data_state -- destroys file state related to data
  */
 void
-vinode_destroy_data_state(struct pmemfile_vinode *vinode)
+vinode_destroy_data_state(PMEMfilepool *pfp, struct pmemfile_vinode *vinode)
 {
+	(void) pfp;
```
Comment (marcinslusarz) [inline src/libpmemfile-core/data.c]: Why? Are you going to use it in a future patch?

Comment (marcinslusarz) [inline src/libpmemfile-core/data.c]: LOL, now I understand. Please add a comment explaining that.

Comment (GBuella) [inline src/libpmemfile-core/data.c]: Done

File: src/libpmemfile-core/data.c [hunk 2]
```diff
@@ -916,31 +921,178 @@ pmemfile_pwrite(PMEMfilepool *pfp, PMEMfile *file, const void *buf,
...
+	while (block != NULL && block->offset + block->size > offset) {
+		if (is_block_contained_by_interval(block, offset, len)) {
+			/*
+			 * Deallocate the whole block, if it is wholly contained
+			 * by the specified interval.
+			 *
+			 *   offset                          offset + len
+			 *   |                                |
+			 * --+-------+-------+----------------+-----
+			 *           | block |
```
Comment (marcinslusarz) [inline src/libpmemfile-core/data.c]: BTW, those ascii art images were very helpful. Thanks for drawing them :)

File: src/libpmemfile-core/data.c [hunk 3]
```diff
@@ -916,31 +918,178 @@ pmemfile_pwrite(PMEMfilepool *pfp, PMEMfile *file, const void *buf,
...
+			 *                                +---+---+
+			 *                                |   |
+			 *                                +---+
+			 *                                 intersection
+			 */
+
+			if (is_block_data_initialized(block)) {
+				uint64_t bl_end = block->offset + block->size;
+				uint64_t in_end = offset + len;
+				uint64_t zero_len = bl_end - in_end;
```
Comment (GBuella) [inline src/libpmemfile-core/data.c]: Right, fixed.

**Comment by marcinslusarz** [comment]
LGTM

**Comment by marcinslusarz** [comment]
Reviewed 4 of 5 files at r1, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/marcinslusarz/pmemfile/41)*

**Pull request merged**